### PR TITLE
Added support for pthread_cancel handling.

### LINKED ISSUE #1171: pthread_cancel not working
In pthread_cancel it will get the tid from pthread_t structure, but the tid field has been changed to the virtual tid in thread_start=>procssChildThread=>dmtcp_init_virtual_tid, then tgkill syscall will fail in such case.

image

Please note it still works in dtmcp(2.6.0) version, as you can see the picture above with strace result, "p.1" is the working case while "p.2" is not working.

### PR DESCRIPTION
**Pull request by karya0**
Libc uses pthread_kill internally to signal to-be-canceled thread. This call fails because pthread_kill uses virtual tid stored in the pthread descriptor. We get around this issue by making an explicit call to pthread_kill _after_ libc has processed the rest of the pthread_cancel call. This is based on the observation (i.e. libc:pthread_cancel implementation) that pthread_kill is the last step in pthread_cancel implementation.

While fixing this issue, found two more long-standing corner cases in virtual pid table implementation that are fixed here as well. A third issue in virtual id implementation is left as is, albeit with a todo.

Added a new test case which fails without this fix and passes with this fix.

Fixes #1171

### CODE REVIEW

File: test/pthread6.c [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+void *
+start_routine(void *arg)
+{
+  int threadId = *(int*)arg;
+  free(arg);
+  printf("ThreadId: %d alive\n", threadId);
+
+  sem_post(&threadInitialized);
+
+  // Will get canceled in here.
```
Comment (gc00) [inline test/pthread6.c:21]: Suggest adding to the above comment:
```
// Will get canceled in here.  'sleep()' is a cancellation point, checked at end of sleep.
```

File: test/pthread6.c [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+    int res = pthread_create(&thread, &attr, start_routine, arg);
+    if (res != 0) {
+      fprintf(stderr, "error creating thread: %s\n", strerror(res));
+      return -1;
+    }
+
+    // Wait for thread to finish initialization.
+    sem_wait(&threadInitialized);
+
+    res = pthread_cancel(thread);
```
Comment (gc00) [inline test/pthread6.c:53]: ADD COMMENT:
```
// We could do a sem_post() here, and then do a sem_wait() before 'sleep(10)'
// in the other thread, in order to guarantee that pthread_cancel()
// executes before sleep(10) completes.  But that would defeat
// the purpose of this test.  We want to execute 'pthread_cancel()' before
// checkpoint and then to have 'sleep(10)' finish executing after restart. 
```

**Review by gc00 — APPROVED** [review]
Two suggested comments to add for clarity.
But I'm approving this PR in advance.
